server/lib/chat/chat-models.ts

```diff
@@ -578,3 +580,116 @@ export async function findChannelByName(
     done()
   }
 }
+
+/**
+ * Returns a list of non-private chat channels, optionally filtered by a `searchStr`. The list
+ * doesn't include the user's joined channel by default, but can be configured to include them as
+ * well.
+ */
+export async function listChannels(
+  {
+    userId,
+    limit,
+    pageNumber,
+    searchStr,
+    joined,
+  }: {
+    /** User requesting the list of channels. */
+    userId: SbUserId
+    /** Amount of channels to return. */
+    limit: number
+    /** Offset from which to start listing the channels. */
+    pageNumber: number
+    /** String to filter the channels by. */
+    searchStr?: string
+    /** Flag which includes the user's joined channels if true. */
+    joined?: boolean
+  },
+  withClient?: DbClient,
+): Promise<ChannelInfo[]> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = joined
+      ? sql`
+        SELECT c.*
+        FROM channels c
+      `
+      : sql`
+        WITH unjoined_channels AS (
+          SELECT DISTINCT(channel_id)
+          FROM channel_users
+          WHERE channel_id NOT IN (SELECT channel_id FROM channel_users WHERE user_id = ${userId})
+        )
+        SELECT c.*
+        FROM channels c
+        INNER JOIN unjoined_channels ON c.id = unjoined_channels.channel_id
+      `
+
+    const whereCondition = sql`WHERE private = false`
+    if (searchStr) {
+      const escapedStr = `%${escapeSearchString(searchStr)}%`
+      whereCondition.append(sql` AND name ILIKE ${escapedStr}`)
+
+      query.append(whereCondition)
+    }
+
+    query.append(sql`
+      GROUP BY c.id
+      ORDER BY c.user_count DESC, c.name
+      LIMIT ${limit}
+      OFFSET ${pageNumber * limit}
+    `)
+
+    const result = await client.query<DbChannel>(query)
+
+    return result.rows.map(row => convertChannelFromDb(row))
+  } finally {
+    done()
+  }
+}
+
+/**
+ * Returns a full list of chat channels, optionally filtered by a `searchStr`. Only admins should be
+ * able to call this function.
+ */
+export async function listAllChannels(
+  {
+    limit,
+    pageNumber,
+    searchStr,
+  }: {
+    limit: number
+    pageNumber: number
+    searchStr?: string
+  },
+  withClient?: DbClient,
+): Promise<ChannelInfo[]> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = sql`
+      SELECT *
+      FROM channels
+    `
+
+    let whereCondition: SQLStatement | undefined
```

Comment:
Hmm, this actually produces a syntax error in SQL for some reason:

![image](https://user-images.githubusercontent.com/4757944/201470916-4378b848-020a-4e2d-9879-2979806bd316.png)

Adding quotes around the whole search expression gets rid of the syntax error, but produces a different error:

![image](https://user-images.githubusercontent.com/4757944/201470965-582e3fe7-3f42-4d72-a464-fbedd09b9ab9.png)

This works though for some reason 🤷‍♂️:
```js
const escapedStr = `%${escapeSearchString(searchStr)}%`
query.append(sql`WHERE name ILIKE ${escapedStr}`)
```

Reply:
That's just because of how the library is choosing to pass in the string, I think. In my code, it passes the string separate from the `%`s, in yours it concatenates them before passing them as a single parameter. So you could fix it in a single line by doing:
```ts
query.append(sql`WHERE name ILIKE ${`%${escapeSearchString(searchStr)}%`}`)
```